### PR TITLE
std: add EnumSet type

### PR DESCRIPTION
**Pull request by daurnimator**
Written for the upcoming `.standardOptions`, but independently useful.

I don't love the method names, but I made them match `BufSet` for now.

### CODE REVIEW

File: lib/std/enum_set.zig [hunk 1]
```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2021 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+const std = @import("std.zig");
+const assert = std.debug.assert;
+const testing = std.testing;
+
+pub fn EnumSet(comptime Enum: type) type {
```
Comment (daurnimator) [inline lib/std/enum_set.zig:10]: One thought I had was: should I add a `options` argument for future extensions?

e.g. I can imagine in future wanting to specify endieness for the `PackedIntArray`

File: lib/std/enum_set.zig [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        is_sparse = false;
+    }
+
+    // TODO: when sparse, set up a mapping so that flags is only `flag_count` in size
+
+    return struct {
+        flags: std.PackedIntArray(u1, max - min + 1),
+
+        fn enumToFlagIndex(value: Enum) usize {
+            return @enumToInt(value) + min;
```
Comment (daurnimator) [inline lib/std/enum_set.zig:40]: this probably needs some casting

File: lib/std/enum_set.zig [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        pub fn put(self: *Self, value: Enum) void {
+            self.flags.set(enumToFlagIndex(value), @boolToInt(true));
+        }
+
+        pub fn delete(self: *Self, value: Enum) void {
+            self.flags.set(enumToFlagIndex(value), @boolToInt(false));
+        }
+
+        pub fn exists(self: Self, value: Enum) bool {
```
Comment (jayschwa) [inline lib/std/enum_set.zig:70]: `set.contains(val)` reads a little better than `set.exists(val)` to me.

**Comment by data-man** [comment]
Is it unusable for enums with >65535 fields?

**Comment by daurnimator** [comment]
> Is it unusable for enums with >65535 fields?

it should be fine; is there something that you think wouldn't work?

**Comment by data-man** [comment]
> it should be fine

How is it possible if ints can't contain > 65535 bits?

> is there something that you think wouldn't work?

E.g. Unicode codepoint names.

**Comment by daurnimator** [comment]
> How is it possible if ints can't contain > 65535 bits?

this PR stores the data in a `PackedIntArray`: it uses n/8 u8s per field. i.e. an enum with 65535 fields would result in an `EnumSet` of size `65535/8`==8192 bytes. There's no reason this couldn't scale to the amount of address space available, though In practice, enums rarely get very large. I mainly expect this PR to be used for things like release modes (an enum with 4 elements) or cpu features (up to ~100 elements)

**Comment by data-man** [comment]
The alternative PoC:
```zig
const std = @import("std");
const mem = std.mem;
const meta = std.meta;
const testing = std.testing;

pub fn fieldsToPackedStruct(comptime T: type) type {
    const fields = meta.fields(T);
    var struct_fields: [fields.len]std.builtin.TypeInfo.StructField = undefined;
    @setEvalBranchQuota(10_000);
    inline for (fields) |f, i| {
        struct_fields[i] = std.builtin.TypeInfo.StructField{
            .name = f.name,
            .field_type = bool,
            .default_value = false,
            .is_comptime = false,
            .alignment = @alignOf(bool),
        };
    }

    return @Type(std.builtin.TypeInfo{
        .Struct = std.builtin.TypeInfo.Struct{
            .is_tuple = false,
            .layout = .Packed,
            .decls = &[_]std.builtin.TypeInfo.Declaration{},
            .fields = &struct_fields,
        },
    });
}

test "fieldsToPackedStruct" {
    const E1 = enum {
        A,
        B,
    };

    const E2 = error{A};

    const S1 = struct {
        @"1": u8,
        @"2": u8,
    };

    const U1 = union {
        a: u8,
        b: u8,
        c: u8,
        d: u8,
        e: u8,
        f: u8,
        g: u8,
        h: u8,
        i: u8,
        j: u8,
    };

    const e1f_t = comptime fieldsToPackedStruct(E1);
    const e2f_t = comptime fieldsToPackedStruct(E2);
    const sf_t = comptime fieldsToPackedStruct(S1);
    const uf_t = comptime fieldsToPackedStruct(U1);

    testing.expect(@sizeOf(e1f_t) == 1);
    testing.expect(@sizeOf(e2f_t) == 1);
    testing.expect(@sizeOf(sf_t) == 1);
    testing.expect(@sizeOf(uf_t) == 2);

    testing.expect(meta.fields(e1f_t).len == 2);
    testing.expect(meta.fields(e2f_t).len == 1);
    testing.expect(meta.fields(sf_t).len == 2);
    testing.expect(meta.fields(uf_t).len == 10);

    testing.expect(mem.eql(u8, meta.fields(e1f_t)[0].name, "A"));
    testing.expect(mem.eql(u8, meta.fields(e2f_t)[0].name, "A"));
    testing.expect(mem.eql(u8, meta.fields(sf_t)[0].name, "1"));
    testing.expect(mem.eql(u8, meta.fields(uf_t)[0].name, "a"));

    var e1f = mem.zeroes(e1f_t);
    testing.expect(!e1f.A);
    e1f.A = true;
    testing.expect(e1f.A);

    var uf = mem.zeroes(uf_t);
    testing.expect(!uf.j);
    uf.j = true;
    testing.expect(uf.j);
}
```

**Comment by daurnimator** [comment]
(note: CI failed due to out of memory)

If PR is accepted, I saw another use-site to follow up with: `fields_seen` in std.json

**Review by andrewrk — CHANGES_REQUESTED** [review]
Now that #7954 is merged, can you look into implementing this as a wrapper around StaticBitSet?

**Pull request closed without merging**